hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationCommandCreator.java

```diff
@@ -0,0 +1,239 @@
+import org.apache.hadoop.ozone.protocol.commands.ReplicateContainerCommand;
+import org.apache.hadoop.ozone.protocol.commands.SCMCommand;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Handles the EC Under replication processing and forming the respective SCM
+ * commands.
+ */
+public class ECUnderReplicationCommandCreator
+    implements ReplicationCommandCreator {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(ECUnderReplicationCommandCreator.class);
+  private final PlacementPolicy containerPlacement;
+  private final long currentContainerSize;
+  private final NodeManager nodeManager;
+
+  public ECUnderReplicationCommandCreator(
+      final PlacementPolicy containerPlacement, final ConfigurationSource conf,
+      NodeManager nodeManager) {
+    this.containerPlacement = containerPlacement;
+    this.currentContainerSize = (long) conf
+        .getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+            ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.nodeManager = nodeManager;
+  }
+
+  /**
+   * Identify a new set of datanode(s) to reconstruct the container and form the
+   * SCM command to send it to DN. In the case of decommission, it will just
+   * generate the replicate commands instead of reconstruction commands.
+   *
+   * @param replicas - An instance of ContainerReplicaCount, containing the
+   *                   current replica count and inflight adds and deletes
+   * @param pendingOps - Inflight replications and deletion metrics.
+   * @param result - Health check result.
+   * @param remainingMaintenanceRedundancy - represents that how many nodes go
+   *                                      into maintenance.
+   * @return Returns the key value pair of destination dn where the command gets
+   * executed and the command itself.
+   */
+  @Override
+  public Map<DatanodeDetails, SCMCommand> processAndCreateCommands(
+      final Set<ContainerReplica> replicas,
+      final ContainerReplicaPendingOps pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas,
+            pendingOps.getPendingOps(container.containerID()),
+            remainingMaintenanceRedundancy);
+    ContainerHealthResult containerHealthResult =
+        new ECContainerHealthCheck().checkHealth(container, replicaCount);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+
+    if (containerHealthResult.getHealthState()
```

Comment:
in fact , the the health state of an EC container may be both under and over replicated. for example, EC-3-2, the health stat is 1,2,3,4,5, but we might have 1,1,2,3,4. so , think we should have more enum items  for `HealthState`

Reply:
For now, to keep things simple, we allow a container to be in only 1 state. Under-replicated is more important that over-replicated, so we handle under-replication first, and then when it has recovered, we will handle the over replication.

Reply:
Yeah, it is ok to give underreplication preference over the overreplication. Once underreplication fixed, next interations can pick up overreplications. Later we can have hybrid state to check both.